tests/store/artifact/test_databricks_artifact_repo.py

```diff
@@ -0,0 +1,215 @@
+        assert repo.run_id == expected_run_id
+        repo = get_artifact_repository(
+            'dbfs:/databricks///mlflow-tracking///EXP_ID///RUN_ID///artifacts/')
+        assert repo.run_id == expected_run_id
+        repo = get_artifact_repository(
+            'dbfs:/databricks///mlflow-tracking//EXP_ID//RUN_ID///artifacts//')
+        assert repo.run_id == expected_run_id
+
+    @pytest.mark.parametrize("artifact_path,expected_location", [
+        (None, 'test.txt'),
+        ('output', 'output/test.txt'),
+        ('', 'test.txt'),
+    ])
+    def test_log_artifact_azure(self, databricks_artifact_repo, test_file, artifact_path,
+                                expected_location):
+        with mock.patch(DATABRICKS_ARTIFACT_REPOSITORY + '._get_write_credentials') \
+                as write_credentials_mock, \
+                mock.patch(DATABRICKS_ARTIFACT_REPOSITORY + '._azure_upload_file') \
+                as azure_upload_mock:
+            mock_credentials = ArtifactCredentialInfo(signed_uri=MOCK_AZURE_SIGNED_URI,
+                                                      type=ArtifactCredentialType.AZURE_SAS_URI)
+            write_credentials_response_proto = GetCredentialsForWrite.Response(
+                credentials=mock_credentials)
+            write_credentials_mock.return_value = write_credentials_response_proto
+            azure_upload_mock.return_value = None
+            databricks_artifact_repo.log_artifact(test_file.strpath, artifact_path)
+            write_credentials_mock.assert_called_with(MOCK_RUN_ID, expected_location)
+            azure_upload_mock.assert_called_with(mock_credentials, test_file.strpath,
+                                                 expected_location)
+
+    @pytest.mark.parametrize("artifact_path,expected_location", [
+        (None, 'test.txt'),
+    ])
+    def test_log_artifact_aws(self, databricks_artifact_repo, test_file, artifact_path,
+                              expected_location):
+        with mock.patch(DATABRICKS_ARTIFACT_REPOSITORY + '._get_write_credentials') \
+                as write_credentials_mock, \
+                mock.patch(DATABRICKS_ARTIFACT_REPOSITORY + '._aws_upload_file') \
+                as aws_upload_mock:
+            mock_credentials = ArtifactCredentialInfo(signed_uri=MOCK_AWS_SIGNED_URI,
+                                                      type=ArtifactCredentialType.AWS_PRESIGNED_URL)
+            write_credentials_response_proto = GetCredentialsForWrite.Response(
+                credentials=mock_credentials)
+            write_credentials_mock.return_value = write_credentials_response_proto
+            aws_upload_mock.return_value = None
+            databricks_artifact_repo.log_artifact(test_file.strpath, artifact_path)
+            write_credentials_mock.assert_called_with(MOCK_RUN_ID, expected_location)
+            aws_upload_mock.assert_called_with(mock_credentials, test_file.strpath)
```

Comment:
@arjundc-db Is it possible to dig into the `_aws_upload_file` method a little further and assert that the expected headers are passed to `requests.put`? This would be a great sanity check that we're correctly handling headers in the client. I think it would be great to do this with a separate test case such as `test_log_artifact_aws_with_headers`.

A similar `test_log_artifact_azure_with_headers` case would also be awesome. Hopefully this doesn't require too much mocking - presumably, just `requests.put` and a couple of `BlobClient` operations for Azure; let me know if you can give this a shot!

Reply:
Done! wew